<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started with NMFscape • NMFscape</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with NMFscape">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">NMFscape</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/NMFscape.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting-started.html">Getting Started</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MicTott/NMFscape" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting Started with NMFscape</h1>
                        <h4 data-toc-skip class="author">Michael
Totty</h4>
            
            <h4 data-toc-skip class="date">2026-01-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/MicTott/NMFscape/blob/HEAD/vignettes/getting-started.Rmd" class="external-link"><code>vignettes/getting-started.Rmd</code></a></small>
      <div class="d-none name"><code>getting-started.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>NMFscape provides fast, scalable non-negative matrix factorization
(NMF) for single-cell and spatial transcriptomic data. NMF decomposes
gene expression data into biologically interpretable gene expression
programs (also called “factors” or “metagenes”) and their weighted
patterns across cells.</p>
<div class="section level3">
<h3 id="what-is-nmf">What is NMF?<a class="anchor" aria-label="anchor" href="#what-is-nmf"></a>
</h3>
<p>Non-negative matrix factorization decomposes a gene expression matrix
<strong>X</strong> (genes × cells) into two lower-rank matrices:</p>
<ul>
<li>
<strong>W</strong> (genes × k programs): Gene loadings - which genes
contribute to each program</li>
<li>
<strong>H</strong> (k programs × cells): Cell loadings - how much
each cell uses each program</li>
</ul>
<p>The key constraint is that all values in W and H are non-negative
(≥0), which enforces a “parts-based” representation where programs
combine additively. This is particularly suited for biological data
where gene expression counts are inherently non-negative.</p>
</div>
<div class="section level3">
<h3 id="why-nmf-for-single-cell-analysis">Why NMF for single-cell analysis?<a class="anchor" aria-label="anchor" href="#why-nmf-for-single-cell-analysis"></a>
</h3>
<ul>
<li>
<strong>Gene program decomposition</strong>: Identifies gene modules
that work together in biological processes</li>
<li>
<strong>Cell type discovery</strong>: Programs often correspond to
cell type-specific signatures, cell states, or biological pathways</li>
<li>
<strong>Biologically interpretable</strong>: Gene programs represent
co-expressed gene modules which can be used downstream for functional
enrichment or pathway analysis</li>
</ul>
<p>NMFscape uses RcppML for high-performance computation and integrates
seamlessly with Bioconductor workflows, storing results directly in
SingleCellExperiment and SpatialExperiment objects.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MicTott/NMFscape" class="external-link">NMFscape</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; Loading required package: generics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'generics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,</span></span>
<span><span class="co">#&gt;     setequal, union</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span><span class="co">#&gt;     rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,</span></span>
<span><span class="co">#&gt;     unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: Seqinfo</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="co">#&gt; Loading required package: SpatialExperiment</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Matrix'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">427</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="basic-nmf-workflow">Basic NMF Workflow<a class="anchor" aria-label="anchor" href="#basic-nmf-workflow"></a>
</h2>
<div class="section level3">
<h3 id="load-and-prepare-data">Load and prepare data<a class="anchor" aria-label="anchor" href="#load-and-prepare-data"></a>
</h3>
<p>We’ll use the Zeisel brain dataset (Zeisel et al. 2015, Science)
which contains ~3,000 cells from the mouse brain with diverse cell types
including neurons, astrocytes, oligodendrocytes, and immune cells. We
remove mitochondrial genes as they can dominate gene programs and are
typically reflective of technical noise rather than biological signals
of interest.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load brain dataset</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html" class="external-link">ZeiselBrainData</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove mitochondrial genes </span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^mt-"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Lognormalize counts</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate PCA and UMAP for visualization (independent of NMF)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-nmf">Run NMF<a class="anchor" aria-label="anchor" href="#run-nmf"></a>
</h3>
<p>Here we run NMF with k=10 programs. Choosing k is non-trivial and
should be chosen with care. It is best practice to use cross-validation
methods to hepl find the optimal k for each dataset (functionaity coming
soon). For now we recommend setting the K between 10-50 for single-cell
datasets.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run NMF with 10 programs</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runNMFscape.html">runNMFscape</a></span><span class="op">(</span><span class="va">sce</span>, k <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># NMF loadings are stored in the reducedDims slot</span></span>
<span><span class="va">nmf_coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"NMF"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">nmf_coords</span><span class="op">)</span>  </span>
<span><span class="co">#&gt; [1] 3005   10</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-nmf-results">Visualizing NMF Results<a class="anchor" aria-label="anchor" href="#visualizing-nmf-results"></a>
</h2>
<div class="section level3">
<h3 id="plot-nmf-programs-on-umap">Plot NMF programs on UMAP<a class="anchor" aria-label="anchor" href="#plot-nmf-programs-on-umap"></a>
</h3>
<p>One key uses of NMF is the ability to find gene program associated
with specific cell types and/or biological states. Cell type-specific
program can easily be identified by plotting program weights within UMAP
gene expression space.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Plot cell type annotations from Zeisel et al.</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"UMAP"</span>, colour_by<span class="op">=</span><span class="st">"level1class"</span>, point_size<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Zeisel Brain Cell Types"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the cell type weights of NMF Program 1</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vizUMAP.html">vizUMAP</a></span><span class="op">(</span><span class="va">sce</span>, program <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"NMF Program 1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display side by side for comparison</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/visualize-programs-1.png" class="r-plt" alt="" width="1152"></p>
</div>
<div class="section level3">
<h3 id="finding-cell-type-specific-programs">Finding cell type specific programs<a class="anchor" aria-label="anchor" href="#finding-cell-type-specific-programs"></a>
</h3>
<p>To help identify programs associated with specific cell types, we can
use dot plots to summarize program usage across annotated cell
types.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># dot plot of NMF program weights aggregated with cell types</span></span>
<span><span class="fu"><a href="../reference/plotProgramDots.html">plotProgramDots</a></span><span class="op">(</span><span class="va">sce</span>, group <span class="op">=</span> <span class="st">"level1class"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/dot-plots-1.png" class="r-plt" alt="" width="576"></p>
</div>
<div class="section level3">
<h3 id="examine-top-weighted-genes-per-program">Examine top weighted genes per program<a class="anchor" aria-label="anchor" href="#examine-top-weighted-genes-per-program"></a>
</h3>
<p>We can examine the top genes contributing to each program using a
heatmap of gene weights from the basis matrix (W). This allows us to see
which genes are most important for defining each program.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Heatmap of top genes per program by weights</span></span>
<span><span class="fu"><a href="../reference/plotProgramHeatmap.html">plotProgramHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, n_genes <span class="op">=</span> <span class="fl">10</span>, cluster_rows<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/gene-weights-1.png" class="r-plt" alt="" width="384"></p>
<p>We can also compare this to the expression of these top genes across
cell types and see that the NMF programs are indeed finding highly
specific marker genes of distinct cell types.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the top 10 genes per program</span></span>
<span><span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopFeatures.html">getTopFeatures</a></span><span class="op">(</span><span class="va">sce</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create list of top 5 unique genes across all programs</span></span>
<span><span class="va">top5_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">top_genes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot heatmap of top 5 genes across all cell types</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, features <span class="op">=</span> <span class="va">top5_genes</span>,</span>
<span>                       group <span class="op">=</span> <span class="st">"level1class"</span>,</span>
<span>                       assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                       scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       center <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>                   <span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/gene-expression-1.png" class="r-plt" alt="" width="384"></p>
</div>
</div>
<div class="section level2">
<h2 id="finding-differentially-expressed-programs">Finding Differentially Expressed Programs<a class="anchor" aria-label="anchor" href="#finding-differentially-expressed-programs"></a>
</h2>
<div class="section level3">
<h3 id="differential-program-usage">Differential Program Usage<a class="anchor" aria-label="anchor" href="#differential-program-usage"></a>
</h3>
<p>To identify which NMF programs are enriched in specific cell types,
we can perform differential expression analysis treating each program as
a “feature” and comparing its usage across cell types. The
<code><a href="../reference/FindAllDEPs.html">FindAllDEPs()</a></code> function performs 1-vs-all comparisons for
each cell type to identify programs that are significantly enriched or
depleted.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find differentially expressed programs across cell types</span></span>
<span><span class="va">deps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindAllDEPs.html">FindAllDEPs</a></span><span class="op">(</span><span class="va">sce</span>, cell_type_col <span class="op">=</span> <span class="st">"level1class"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View results for pyramidal CA1 neurons</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deps</span><span class="op">[[</span><span class="st">"pyramidal CA1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 12 columns</span></span>
<span><span class="co">#&gt;        fold_enrichment mean_usage_in mean_usage_out      p.value          FDR</span></span>
<span><span class="co">#&gt;              &lt;numeric&gt;     &lt;numeric&gt;      &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; NMF_1       17.8345991   9.48009e-04    5.31556e-05 7.27966e-294 7.27966e-293</span></span>
<span><span class="co">#&gt; NMF_3        0.0799193   3.73273e-05    4.67062e-04 2.13475e-280 1.06737e-279</span></span>
<span><span class="co">#&gt; NMF_4        0.2728305   1.17489e-04    4.30628e-04 1.63575e-178 5.45251e-178</span></span>
<span><span class="co">#&gt; NMF_5        0.4988523   1.96831e-04    3.94567e-04 3.05011e-146 7.62528e-146</span></span>
<span><span class="co">#&gt; NMF_8        0.1522181   6.89102e-05    4.52707e-04 4.42411e-127 8.84822e-127</span></span>
<span><span class="co">#&gt; NMF_10       0.2709636   1.16773e-04    4.30954e-04 4.10380e-126 6.83966e-126</span></span>
<span><span class="co">#&gt;        summary.AUC AUC.astrocytes_ependymal AUC.endothelial-mural</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt;                &lt;numeric&gt;             &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; NMF_1   0.99804411               0.99695725              0.998466</span></span>
<span><span class="co">#&gt; NMF_3   0.01386659               0.30947627              0.384918</span></span>
<span><span class="co">#&gt; NMF_4   0.00873590               0.66918644              0.585449</span></span>
<span><span class="co">#&gt; NMF_5   0.00187287               0.66666667              0.710466</span></span>
<span><span class="co">#&gt; NMF_8   0.00784900               0.35272611              0.007849</span></span>
<span><span class="co">#&gt; NMF_10  0.00564335               0.00564335              0.377618</span></span>
<span><span class="co">#&gt;        AUC.interneurons AUC.microglia AUC.oligodendrocytes AUC.pyramidal SS</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt;     &lt;numeric&gt;            &lt;numeric&gt;        &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; NMF_1        0.99692446    0.99829932            0.9980441        0.9917339</span></span>
<span><span class="co">#&gt; NMF_3        0.47545077    0.14485666            0.0138666        0.4002632</span></span>
<span><span class="co">#&gt; NMF_4        0.43132827    0.65901632            0.6902543        0.0087359</span></span>
<span><span class="co">#&gt; NMF_5        0.00187287    0.64696486            0.7117964        0.4785072</span></span>
<span><span class="co">#&gt; NMF_8        0.44007565    0.00035861            0.3561898        0.4324976</span></span>
<span><span class="co">#&gt; NMF_10       0.47463185    0.18815609            0.3689479        0.4898255</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-differential-program-usage">Visualizing Differential Program Usage<a class="anchor" aria-label="anchor" href="#visualizing-differential-program-usage"></a>
</h3>
<p>We can visualize these results using volcano plots via the
<code>plotDEPsVolcano</code> function, which show the relationship
between fold enrichment and statistical significance. It should be noted
that the p-values here are inflated and should be not
over-interpreted.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Volcano plots for all cell types</span></span>
<span><span class="fu"><a href="../reference/plotDEPsVolcano.html">plotDEPsVolcano</a></span><span class="op">(</span><span class="va">deps</span>, fold_threshold <span class="op">=</span> <span class="fl">1.5</span>, pval_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/dep-volcano-1.png" class="r-plt" alt="" width="960"></p>
<p>We can also summarize differential program usage across all cell
types using a heatmap via the <code>plotDEPsHeatmap</code> function to
make visualization easy. We can set the <code>logFC_threshold</code>
parameter (2 by default) to highlight programs with strong cell
type-specificity.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Heatmap of log2 fold change with stars for significant enrichments</span></span>
<span><span class="fu"><a href="../reference/plotDEPsHeatmap.html">plotDEPsHeatmap</a></span><span class="op">(</span><span class="va">deps</span>, logfc_threshold <span class="op">=</span> <span class="fl">2</span>, star_size<span class="op">=</span><span class="fl">14</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/dep-heatmap-1.png" class="r-plt" alt="" width="288"></p>
</div>
<div class="section level3">
<h3 id="downstream-analyses">Downstream analyses<a class="anchor" aria-label="anchor" href="#downstream-analyses"></a>
</h3>
<p>NMF programs can be used for a variety of downstream analyses
including (vingettes coming soon):</p>
<ul>
<li>
<strong>Integration with spatial data</strong>: Map NMF programs
onto spatial transcriptomics data to predict the location of distinct
cell type.</li>
<li>
<strong>Identification of disease-associated gene programs</strong>:
Compare NMF programs between healthy and diseased samples (or case vs
control) to identify programs associated with disease states.</li>
<li>
<strong>Functional enrichment</strong>: Perform GO or pathway
enrichment analysis on top genes from each program to identify
biological processes represented by each program. This is particularly
useful for case-control experiments or disease studies.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 Patched (2026-01-07 r89305)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sequoia 15.6.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] pheatmap_1.0.13             patchwork_1.3.2            </span></span>
<span><span class="co">#&gt;  [3] scRNAseq_2.24.0             scran_1.38.0               </span></span>
<span><span class="co">#&gt;  [5] scater_1.38.0               ggplot2_4.0.1              </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.20.0              Matrix_1.7-4               </span></span>
<span><span class="co">#&gt;  [9] NMFscape_0.99.0             SpatialExperiment_1.20.0   </span></span>
<span><span class="co">#&gt; [11] SingleCellExperiment_1.32.0 SummarizedExperiment_1.40.0</span></span>
<span><span class="co">#&gt; [13] Biobase_2.70.0              GenomicRanges_1.62.1       </span></span>
<span><span class="co">#&gt; [15] Seqinfo_1.0.0               IRanges_2.44.0             </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.48.0            MatrixGenerics_1.22.0      </span></span>
<span><span class="co">#&gt; [19] matrixStats_1.5.0           BiocGenerics_0.56.0        </span></span>
<span><span class="co">#&gt; [21] generics_0.1.4              BiocStyle_2.38.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] BiocIO_1.20.0            bitops_1.0-9             filelock_1.0.3          </span></span>
<span><span class="co">#&gt;   [4] tibble_3.3.1             XML_3.99-0.20            lifecycle_1.0.5         </span></span>
<span><span class="co">#&gt;   [7] httr2_1.2.2              edgeR_4.8.2              lattice_0.22-7          </span></span>
<span><span class="co">#&gt;  [10] pals_1.10                ensembldb_2.34.0         alabaster.base_1.10.0   </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.4           limma_3.66.0             sass_0.4.10             </span></span>
<span><span class="co">#&gt;  [16] rmarkdown_2.30           jquerylib_0.1.4          yaml_2.3.12             </span></span>
<span><span class="co">#&gt;  [19] metapod_1.18.0           otel_0.2.0               mapproj_1.2.12          </span></span>
<span><span class="co">#&gt;  [22] DBI_1.2.3                RColorBrewer_1.1-3       maps_3.4.3              </span></span>
<span><span class="co">#&gt;  [25] abind_1.4-8              AnnotationFilter_1.34.0  RCurl_1.98-1.17         </span></span>
<span><span class="co">#&gt;  [28] rappdirs_0.3.4           ggrepel_0.9.6            irlba_2.3.5.1           </span></span>
<span><span class="co">#&gt;  [31] alabaster.sce_1.10.0     RSpectra_0.16-2          dqrng_0.4.1             </span></span>
<span><span class="co">#&gt;  [34] pkgdown_2.2.0            codetools_0.2-20         DelayedArray_0.36.0     </span></span>
<span><span class="co">#&gt;  [37] tidyselect_1.2.1         UCSC.utils_1.6.1         farver_2.1.2            </span></span>
<span><span class="co">#&gt;  [40] ScaledMatrix_1.18.0      viridis_0.6.5            BiocFileCache_3.0.0     </span></span>
<span><span class="co">#&gt;  [43] GenomicAlignments_1.46.0 jsonlite_2.0.0           BiocNeighbors_2.4.0     </span></span>
<span><span class="co">#&gt;  [46] systemfonts_1.3.1        tools_4.5.2              ragg_1.5.0              </span></span>
<span><span class="co">#&gt;  [49] Rcpp_1.1.1               glue_1.8.0               gridExtra_2.3           </span></span>
<span><span class="co">#&gt;  [52] SparseArray_1.10.8       xfun_0.56                GenomeInfoDb_1.46.2     </span></span>
<span><span class="co">#&gt;  [55] dplyr_1.1.4              HDF5Array_1.38.0         gypsum_1.6.0            </span></span>
<span><span class="co">#&gt;  [58] withr_3.0.2              BiocManager_1.30.27      fastmap_1.2.0           </span></span>
<span><span class="co">#&gt;  [61] rhdf5filters_1.22.0      bluster_1.20.0           RcppML_0.3.7            </span></span>
<span><span class="co">#&gt;  [64] digest_0.6.39            rsvd_1.0.5               R6_2.6.1                </span></span>
<span><span class="co">#&gt;  [67] colorspace_2.1-2         textshaping_1.0.4        dichromat_2.0-0.1       </span></span>
<span><span class="co">#&gt;  [70] RSQLite_2.4.5            cigarillo_1.0.0          h5mread_1.2.1           </span></span>
<span><span class="co">#&gt;  [73] rtracklayer_1.70.1       FNN_1.1.4.1              httr_1.4.7              </span></span>
<span><span class="co">#&gt;  [76] htmlwidgets_1.6.4        S4Arrays_1.10.1          uwot_0.2.4              </span></span>
<span><span class="co">#&gt;  [79] pkgconfig_2.0.3          gtable_0.3.6             blob_1.3.0              </span></span>
<span><span class="co">#&gt;  [82] S7_0.2.1                 XVector_0.50.0           htmltools_0.5.9         </span></span>
<span><span class="co">#&gt;  [85] bookdown_0.46            ProtGenerics_1.42.0      scales_1.4.0            </span></span>
<span><span class="co">#&gt;  [88] alabaster.matrix_1.10.0  png_0.1-8                knitr_1.51              </span></span>
<span><span class="co">#&gt;  [91] rstudioapi_0.18.0        rjson_0.2.23             curl_7.0.0              </span></span>
<span><span class="co">#&gt;  [94] cachem_1.1.0             rhdf5_2.54.1             BiocVersion_3.22.0      </span></span>
<span><span class="co">#&gt;  [97] parallel_4.5.2           vipor_0.4.7              AnnotationDbi_1.72.0    </span></span>
<span><span class="co">#&gt; [100] restfulr_0.0.16          desc_1.4.3               pillar_1.11.1           </span></span>
<span><span class="co">#&gt; [103] grid_4.5.2               alabaster.schemas_1.10.0 vctrs_0.7.0             </span></span>
<span><span class="co">#&gt; [106] BiocSingular_1.26.0      dbplyr_2.5.1             beachmat_2.26.0         </span></span>
<span><span class="co">#&gt; [109] cluster_2.1.8.1          beeswarm_0.4.0           evaluate_1.0.5          </span></span>
<span><span class="co">#&gt; [112] GenomicFeatures_1.62.0   magick_2.9.0             cli_3.6.5               </span></span>
<span><span class="co">#&gt; [115] locfit_1.5-9.12          compiler_4.5.2           Rsamtools_2.26.0        </span></span>
<span><span class="co">#&gt; [118] rlang_1.1.7              crayon_1.5.3             labeling_0.4.3          </span></span>
<span><span class="co">#&gt; [121] fs_1.6.6                 ggbeeswarm_0.7.3         viridisLite_0.4.2       </span></span>
<span><span class="co">#&gt; [124] alabaster.se_1.10.0      BiocParallel_1.44.0      Biostrings_2.78.0       </span></span>
<span><span class="co">#&gt; [127] lazyeval_0.2.2           ExperimentHub_3.0.0      bit64_4.6.0-1           </span></span>
<span><span class="co">#&gt; [130] Rhdf5lib_1.32.0          KEGGREST_1.50.0          statmod_1.5.1           </span></span>
<span><span class="co">#&gt; [133] alabaster.ranges_1.10.0  AnnotationHub_4.0.0      igraph_2.2.1            </span></span>
<span><span class="co">#&gt; [136] memoise_2.0.1            bslib_0.9.0              bit_4.6.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/MicTott" class="external-link">Michael Totty</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
